flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/logical/FlinkLogicalUpsertToRetraction.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.nodes.logical
+
+import java.util.{List => JList}
+
+import org.apache.calcite.plan._
+import org.apache.calcite.rel.{RelNode, SingleRel}
+import org.apache.calcite.rel.convert.ConverterRule
+import org.apache.calcite.rel.metadata.RelMetadataQuery
+import org.apache.flink.table.plan.logical.rel.LogicalUpsertToRetraction
+import org.apache.flink.table.plan.nodes.FlinkConventions
+
+class FlinkLogicalUpsertToRetraction(
```

Comment:
Only calc don't remove key fields can be pushed down. We make sure this in `CalcUpsertToRetractionTransposeRule`. After optimization, keyNames will be renames according to the calc.

Reply:
`FlinkLogicalUpsertToRetraction` has hidden `RexNode` reference - in form of `keyNames`.

I think currently calcite is only using `accept()` or `getChildExps()` for decorrelating the plan. So currently there is no problem, because `FlinkLogicalUpsertToRetraction.keyNames` should never have anything to do with corellated subqueries. But my point is that if someone would use those methods for something else, like renaming variables, looking which fields are referenced, etc..., it would brake very silently and the failure would be very hard to debug.

Look at the all of the implementations of `org.apache.calcite.rel.RelNode#accept(org.apache.calcite.rex.RexShuttle)`. Basically every node that has some some input references (or other expressions), is implementing this method.

Unfortunately you might be right, that throwing an exception will not work here :/

Reply:
Thanks for your reply. I guess we are on the same page now?

Reply:
I'm syncing with Calcite dev mailing list, trying to understand those methods. Until we get a clear response from them lets leave this as it is now.

Reply:
Thank you very much for looking into this. I will also keep an eye on the Calcite dev mailing list.

Reply:
They have responded to me and basically there is no contract... as long as the tests are passing we are fine. I don't like it, but if you want we can leave those methods not implemented.